Docker compose

markrcote opened the pull request:
Support for docker-compose for local development (not testing yet).

markrcote commented:
Updated once more now that I realized that I could use a standard rabbitmq docker image. And I added persistence to the db service.

edmorley commented on docker-compose.yml:
```diff
@@ -0,0 +1,37 @@
+version: "2"
+services:
+  db:
+    image: postgres
```
This is equivalent to `postgres:latest` which is currently v10.1 but will change over time. I'd recommend pinning to a major version:
https://hub.docker.com/_/postgres/

Also the alpine image is only 14MB vs the 103MB of the standard image - so possibly worth trying that for faster CI and local development bootstrap times.

edmorley commented on docker-compose.yml:
```diff
@@ -0,0 +1,37 @@
+    ports:
+      - "5432:5432"
+  pulse:
+    image: rabbitmq:3-management
```
Standard vs alpine is 76MB vs 44MB:
https://hub.docker.com/r/library/rabbitmq/tags/

edmorley commented on docker/Dockerfile-pgbase:
```diff
@@ -0,0 +1,15 @@
+FROM python:2.7
```
I'd use `python:2.7-slim` (or even better, `python:2.7-slim-stretch`, which is based on the newer "Debian Stretch") - since they are 50MB compared to the standard image's 273MB (!).

edmorley commented on docker/Dockerfile-pgbase:
```diff
@@ -0,0 +1,15 @@
+RUN apt-get update && apt-get install -y \
+    libffi-dev \
+    libssl-dev \
+    postgresql-client
```
I'd add `rm -rf /var/lib/apt/lists/*` here to clean up after the apt-get update - like the official images do:
https://github.com/docker-library/python/blob/a4ce69de053337ec22d7f322cd997398eeba2350/2.7/stretch/slim/Dockerfile#L17

edmorley commented on docker/Dockerfile-pgbase:
```diff
@@ -0,0 +1,15 @@
+ENV DATABASE_URL postgres://postgres:postgres@db:5432/postgres
+ENV FLASK_SECRET_KEY GBhQrZfbN9KGMcyLVImGeS2Wtj3SFCdF
+ENV RABBIT_MANAGEMENT_URL http://pulse:15672/api/
+RUN mkdir /code
```
Nit: The `mkdir` can be removed since the `WORKDIR` command will create the directory automatically:
https://docs.docker.com/engine/reference/builder/#workdir

edmorley commented on docker/Dockerfile-pgbase:
```diff
@@ -0,0 +1,15 @@
+ENV RABBIT_MANAGEMENT_URL http://pulse:15672/api/
+RUN mkdir /code
+WORKDIR /code
+ADD requirements.txt /code/
```
Nit: For these simpler cases, whilst `ADD` has the same effect as `COPY`, the latter is preferred for clarity:
https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#add-or-copy

edmorley commented on docker/Dockerfile-pgbase:
```diff
@@ -0,0 +1,15 @@
+RUN mkdir /code
+WORKDIR /code
+ADD requirements.txt /code/
+RUN pip install --no-cache-dir -r requirements.txt
```
Great spot on use of `--no-cache-dir` :-)

edmorley commented on docker/Dockerfile-pgbase:
```diff
@@ -0,0 +1,15 @@
+ADD requirements.txt /code/
+RUN pip install --no-cache-dir -r requirements.txt
+ADD . /code/
+RUN python setup.py develop
```
Nit: I think `setup.py develop` is generally discouraged now in favour of `pip install -e .` - though I could be misremembering?

edmorley commented on docker/Dockerfile-pgguardian:
```diff
@@ -0,0 +1,2 @@
+FROM pulseguardian_pulseguardian-base
```
Another way of implementing this file and `docker/Dockerfile-pgweb` would be to move the contents to `docker-compose.yml` as additional entries with differing `environment` and `command` keys. Though subjective style choice so doesn't matter either way :-)

edmorley commented on pulseguardian/model/base.py:
```diff
@@ -22,4 +24,10 @@
+        try:
+            Base.metadata.create_all(bind=engine)
+        except OperationalError:
+            time.sleep(5)
```
What do you think about adding a `log.info()` here saying "waiting for DB" or "Retrying DB connection" or ...?

markrcote commented:
Sigh, I did this in a series of fix-up commits to make it easier to review, which I would later have folded into the appropriate commits, since GitHub sucks at handling amended commits... buuuut I rebased onto master so GH lost the commit mapping anyway. 🤣

I'll just go ahead and amend the original commits.

markrcote commented:
@edmorley Okay this is ready for re-review whenever you have a chance. I addressed all your points except the Dockerfile-dependencies, which as you say is subjective. 😃

markrcote commented on docker/Dockerfile-pgbase:
```diff
@@ -0,0 +1,19 @@
+FROM python:2.7-slim-stretch
+# Work around man directories not being installed in the 'slim' image.
```
Yeah, it breaks the installation of the postgres client, which expects the man directory to be there. I'll update the comment.

edmorley replied:
Shoddy ubuntu stock packages strike again! :-)
(Probably not worth it here, but I know quite a few people use the official APT repo instead: https://www.postgresql.org/download/linux/ubuntu/)
